datastacks/utils.py

```diff
@@ -0,0 +1,97 @@
+from jinja2 import Environment, FileSystemLoader
+from pathlib import Path
+
+
+def generate_target_dir(stage_name: str, dataset_name: str) -> str:
+    """Uses stage name and name of the dataset to generate the target directory to write to.
+
+    Args:
+        stage_name: Name of the pipeline stage eg. Ingest
+        dataset_name: Name of the dataset being processed
+
+    Returns:
+        Path to render template into
+    """
+    target_dir = f"de_workloads/{stage_name}/{stage_name}_{dataset_name}"
+    return target_dir
+
+
+def render_template_components(config: BaseModel, template_source_path: str, target_dir: str) -> None:
+    """Renders all templates within a given path with provided config, and saves results into a new target path,
+    while maintaining folder structure and removing jinja file extensions, any existing files with the same name
+    are overwritten.
+
+    Args:
+        config: Pydantic model of config containing required templating params
+        template_source_path: Path containing templates to be rendered
+        target_dir: Directory to render templates into
+    """
+    Path(target_dir).mkdir(parents=True, exist_ok=True)
+    template_loader = FileSystemLoader(searchpath=str(Path(template_source_path).absolute()))
+    template_env = Environment(loader=template_loader, autoescape=True)
+
+    template_list = template_env.list_templates(extensions=".jinja")
+    for template in template_list:
+        template = template_env.get_template(template)
+        template_filepath = Path(template.filename.split(template_source_path, 1)[1])
+        template_path = template_filepath.parent
+        template_filename = template_filepath.stem
+        Path(target_dir / template_path).mkdir(parents=True, exist_ok=True)
+        template.stream(config).dump(f"{target_dir}/{template_path}/{template_filename}")
+
+
+def generate_pipeline(config_path: str, dq_flag: bool, template_source_folder: str, stage_name: str) -> str:
+    """Reads in config from given file, renders templates for new pipeline,
+    writes out to new path, and returns the target directory it wrote out to.
+    If directory already exists it asks for user input to confirm overwrite.
+
+    Args:
+        config_path: Path to config file containing templating params
+        dq_flag: Flag indicating whether to include data quality components or not
+        template_source_folder: Name of the folder within the template directory
+            containing the templates to be rendered
+        stage_name: Name of the pipeline stage eg. Ingest
+
+    Returns:
+        Path to rendered template
+    """
+    stage_name = stage_name.capitalize()
+
+    click.echo("Reading config from provided path...")
+    with open(config_path, "r") as file:
+        config_dict = yaml.safe_load(file)
+    config = IngestConfig.parse_obj(config_dict)
+
+    click.echo("Successfully read config file.\n")
+
+    dataset_name = config.dataset_name.lower()
```

Comment:
I might be missing something, but when we do lower() here without updating it in the passed config, the pipeline name, which is in the templates determined as:

`{% set pipeline_name = 'Ingest_' ~ dataset_name -%}`

will be potentially named differently from the directory. Or am I wrong?

Reply:
Honestly I was only considering it for the directory.

Reply:
I'll resolve, although I think that enforcing all the capitalization is a separate topic and I'm not sure that it's required here :)

Reply:
Fixed.

Reply:
Okay, well it'll need to be a wider conversation then as Andy has requested it above.